### ISSUE TITLE
TestServer exception handling differs from Kestrel/regular hosted exception handling

### ISSUE THREAD

**Issue by wizofaus** [dotnet/aspnetcore#27208]
Normally if your Controller function throws an exception and there's no middleware etc. in place to catch it, the client will get a 500 internal server error.
But using TestServer they don't - the unit test can actually catch the exception thrown by the controller function. Given the purpose of TestServer is to enable writing tests that emulate a client making actual HTTP calls to the your server, surely the behaviour in this case should be the same?
Accepted it's relatively easy to configure the IApplicationBuilder with UseExceptionHandler( ) in this case but it just seems odd that the default behaviour is the way it is, nor is it something I'd expect we could rely on (e.g. if something was changed in the future that meant TestServer was implemented differently and there wasn't a regular call-stack between the unit test and the controller function that an exception could be propagated back through).

BTW I noticed this because another developer in our team had written a bunch of unit tests that checked that the exceptions thrown by the controller functions were being caught, and I had to explain why that didn't make sense.

**Comment by Tratcher** [comment]
This difference is intentional as TestServer is intended for use only in unit tests. Passing through the original exception makes the tests easier to write and debug. If it only returned a 500 response like Kestrel you'd have a hard time identifying the reason without attaching the debugger.

**Comment by Tratcher** [comment]
In your situation there are two servers/controllers and you're calling from one to the other via HttpClient and TestServer? Ok, that is a situation where a 500 might would make more sense. You'd still have a hard time tracking down the reason for the exception, but your front controller would get the expected response.

This situation is less common than TestServer's standard scenario for directly testing a component, so any change here would have to be opt in.

**Comment by wizofaus** [comment]
Thanks for the quick response!
There's no need to track down a "reason" for the 500 error, the test is deliberately written in such a way that we expect the returned result to be a 500.
As it happens we decided we needed to ensure the unit test setup code installed the same exception handler middleware used at runtime, but it just struck me that it was not behaviour I expected.  If it's by design and intended to be something a developer can rely on when writing tests then OK, though I personally wouldn't want to write a test that way (my philosophy is that the test should work if even if it were a real HttpClient making an actual HTTP socket connection).
